Rover: add 'Loiter or Hold' as a failsafe action

Williangalvani opened the pull request:
This is especially useful for boats, as we usually don't want it to drift away with the current once if failsafes

rmackay9 reviewed (approved):
Hi @Williangalvani,

Thanks very much for this!  It looks good, I just wanted to check that it's been tested in SITL?

rmackay9 commented:
@Williangalvani,

Tell me when you're ready and I think we can merge this

Williangalvani commented:
> @Williangalvani,
> 
> Tell me when you're ready and I think we can merge this

I think it is ready! I just added a test. Can you take a look? There was an older, commented out code, so I just replaced that.

rmackay9 commented on Tools/autotest/rover.py:
```diff
@@ -174,55 +174,38 @@ def drive_left_circuit(self):
+        self.set_parameters({
+            "FS_THR_ENABLE": 1,
+            "FS_ACTION": 6,
+            "FS_THR_VALUE": 1000,
```
Perhaps @peterbarker knows better but for me, instead of setting FS_THR_VALUE and "self.set_rc(3, 900)" I would use the SIM_RC_FAIL parameter

self.set_parameter("SIM_RC_FAIL", 1)

and then to restore RC use:

self.set_parameter("SIM_RC_FAIL", 0)

functionally this makes no difference though

Williangalvani replied:
nice! updated

rmackay9 commented on Tools/autotest/rover.py:
```diff
@@ -174,55 +174,38 @@ def drive_left_circuit(self):
+        self.wait_mode("HOLD", timeout=10)
+        self.progress("Loiter or Hold as throttle failsafe OK")
+
+        self.set_rc(3, 1000)
```
I think you can remove this stuff at the bottom because we have push and restore now before and after each test

Williangalvani replied:
ok, I only needed to keep the disarm() for the autotest to stay happy

peterbarker commented on Rover/Rover.h:
```diff
@@ -416,7 +416,8 @@ class Rover : public AP_Vehicle {
         SmartRTL_Hold = 4,
-        Terminate     = 5
+        Terminate     = 5,
+        Loiter_Hold   = 6
```
```suggestion
        Loiter_Hold   = 6,
```

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -174,55 +174,34 @@ def drive_left_circuit(self):
+        self.arm_vehicle()
+
+        self.progress("Testing failsafe to loiter")
+        self.set_parameter("SIM_RC_FAIL", 1)
```
So this is different to what was here a moment ago.

This was setting throttle to a bind-time value, now you are setting SITL to be "no pulses on RC input" instead.

So you're not really "testing throttle failsafe" here now, you are testing *RC* failsafes.

This is fine - but I *am* planning on renaming these parameters :-)

```suggestion
        self.set_parameter("SIM_RC_FAIL", 1)  # no-pulses failsafe
```

rmackay9 replied:
So for me, throttle failsafe and RC failsafe are the same thing.  They may be triggered by slightly different signals sent by the RC receiver (e.g. futaba receivers send a pulse below 900us, other receivers send no pulses at all)... but they're essentially the same thing

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -174,55 +174,34 @@ def drive_left_circuit(self):
+
+        # restore RC and prepare to test hold fallback
+        self.set_parameter("SIM_RC_FAIL", 0)
+        self.set_parameter("SIM_GPS1_JAM", 1)
```
Usually we would fail the GPSs by disabling rather than jam them.... but this apparently works!

Williangalvani replied:
disabling sounds better. updated.

peterbarker commented on Tools/autotest/rover.py:
```diff
@@ -174,55 +174,34 @@ def drive_left_circuit(self):
+        self.wait_mode("LOITER", timeout=10)
+
+        # restore RC and prepare to test hold fallback
+        self.set_parameter("SIM_RC_FAIL", 0)
```
Maybe use set_parameters here?  It's marginalyl faster

Williangalvani replied:
done!

Williangalvani replied:
actually now that I switched to changing SIM_GPS1_ENABLE, it is passing without changing GPS2. I wonder if that was just flaky because I was using JAM instead

IamPete1 commented on Rover/fence.cpp:
```diff
@@ -67,6 +67,11 @@ void Rover::fence_check()
                         set_mode(mode_hold, ModeReason::FENCE_BREACHED);
                     }
                     break;
+                case FailsafeAction::Loiter_Hold:
```
I think you need the docs for this one up in AC_Fence too.

IamPete1 replied:
https://github.com/ArduPilot/ardupilot/blob/7c0f01bd8d4c335a3fb962cfce925640cc1d4307/libraries/AC_Fence/AC_Fence.cpp#L70

Williangalvani replied:
thanks! that was missing "terminate" too
